models/running_processor.go

```diff
@@ -71,34 +71,43 @@ func (r *RunningProcessor) Init() error {
 	return nil
 }
 
-func (rp *RunningProcessor) Apply(in ...telegraf.Metric) []telegraf.Metric {
-	rp.Lock()
-	defer rp.Unlock()
+func (r *RunningProcessor) ApplyFilters(metric telegraf.Metric) telegraf.Metric {
+	if ok := r.Config.Filter.Select(metric); !ok {
+		return nil // pass downstream
+	}
 
-	ret := []telegraf.Metric{}
+	r.Config.Filter.Modify(metric)
+	if len(metric.FieldList()) == 0 {
+		r.metricFiltered(metric)
+		// caller needs to notice there's no fields and that it needs to be dropped
+		// from the stream
+		return metric
+	}
 
-	for _, metric := range in {
-		// In processors when a filter selects a metric it is sent through the
-		// processor.  Otherwise the metric continues downstream unmodified.
-		if ok := rp.Config.Filter.Select(metric); !ok {
-			ret = append(ret, metric)
+	return metric
+}
+
+func (r *RunningProcessor) Log() telegraf.Logger {
+	return r.log
+}
+
+func (r *RunningProcessor) Start(in <-chan telegraf.Metric, acc telegraf.MetricStreamAccumulator) error {
+	if err := r.Processor.Start(acc); err != nil {
+		return err
+	}
+
+	for m := range in {
+		filteredMetric := r.ApplyFilters(m)
+		if filteredMetric == nil {
+			acc.PassMetric(m)
 			continue
 		}
-
-		rp.Config.Filter.Modify(metric)
-		if len(metric.FieldList()) == 0 {
-			rp.metricFiltered(metric)
+		if len(m.FieldList()) == 0 {
+			acc.DropMetric(filteredMetric)
```

Comment:
Would be best to merge these functions again so you don't need to double check the number of fields.  Either that or return multiple values, but I'm not sure I see the advantage of the change.  Make sure the metric is properly dropped (must call metric.Drop()).